man/dmbackup.1
.TH "DMBACKUP" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmbackup \(en creates an online backup of a running database
.SH SYNOPSIS
\f[B]dmbackup\f[R] \-\-help
.PP
\f[B]dmbackup\f[R] \-\-version
.PP
\f[B]dmbackup\f[R] \-\-\f[B]database\f[R] \f[I]file\f[R]
\-\-\f[B]backup\f[R] \f[I]file\f[R] [\-\-\f[B]vacuum\f[R]]
[\-\-\f[B]wal\f[R]]
.SH DESCRIPTION
The \f[B]dmbackup\f[R] utility creates an online backup of a running
SQLite database.
By default, the SQLite backup API is used.
The program is functional equivalent to running the \f[I]sqlite3(1)\f[R]
command\-line interface:
.IP
.EX
$ sqlite3 <database> \(dq.backup \(aq<backup>\(aq\(dq
.EE
.PP
\f[B]dmbackup\f[R] does not replace existing backup databases.
.SH OPTIONS
.TP
\f[B]\-\-backup\f[R], \f[B]\-b\f[R] \f[I]file\f[R]
Path of the backup database.
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
Path of the SQLite database to backup.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-vacuum\f[R], \f[B]\-U\f[R]
Use \f[CR]VACUUM\f[R] \f[CR]INTO\f[R] instead of the SQLite backup API.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print progress to \f[I]stdout\f[R] (not in vacuum mode).
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.TP
\f[B]\-\-wal\f[R], \f[B]\-W\f[R]
Enable WAL mode for backup database.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Backup has been created.
.TP
\f[B]1\f[R]
Failure.
Backup creation failed.
.SH EXAMPLE
Create an online backup of an observation database:
.IP
.EX
$ dmbackup \-d /var/dmpack/observ.db \-b /tmp/observ.db
.EE
.SH SEE ALSO
\f[I]dminit(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmbackup.f90
! dmbackup.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmbackup
    !! Creates an online backup of an SQLite database.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmbackup'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    integer, parameter :: APP_NSTEPS     = 500 !! Step size for backup API.
    integer, parameter :: APP_SLEEP_TIME = 25  !! Sleep time between steps in msec.

    type :: app_type
        !! Command-line arguments.
        character(FILE_PATH_LEN) :: database = ' '     !! Path to database.
        character(FILE_PATH_LEN) :: backup   = ' '     !! Path to backup.
        logical                  :: vacuum   = .false. !! Vacuum flag.
        logical                  :: wal      = .false. !! WAL flag.
        logical                  :: verbose  = .false. !! Verbose flag.
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Create selected database type.
    rc = backup(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)
contains
    integer function backup(app) result(rc)
        !! Creates database backup.
        type(app_type), intent(inout) :: app

        type(db_type) :: db

        ! Open database.
        rc = dm_db_open(db, app%database)

        if (dm_is_error(rc)) then
            call dm_error_out(rc, 'failed to open database ' // app%database)
            return
        end if

        backup_block: block
            ! Use VACUUM INTO.
            if (app%vacuum) then
                rc = dm_db_vacuum(db, into=app%backup)
                exit backup_block
            end if

            ! Use SQLite backup API.
            if (app%verbose) then
                ! Using callback.
                rc = dm_db_backup(db=db, path=app%backup, wal=app%wal, callback=backup_callback, &
                                  nsteps=APP_NSTEPS, sleep_time=APP_SLEEP_TIME)
                print *
            else
                ! No callback.
                rc = dm_db_backup(db=db, path=app%backup, wal=app%wal, nsteps=APP_NSTEPS, &
                                  sleep_time=APP_SLEEP_TIME)
            end if
        end block backup_block

        call dm_error_out(rc, 'backup failed')
        call dm_db_close(db)
    end function backup

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments.
        type(app_type), intent(out) :: app

        type(arg_parser_class) :: parser

        call parser%add('database', short='d', type=ARG_TYPE_DATABASE, required=.true., exist=.true.) ! -d, --database <path>
        call parser%add('backup',   short='b', type=ARG_TYPE_DATABASE, required=.true.)               ! -b, --backup <path>
        call parser%add('vacuum',   short='U', type=ARG_TYPE_LOGICAL) ! -U, --vacuum
        call parser%add('wal',      short='W', type=ARG_TYPE_LOGICAL) ! -W, --wal
        call parser%add('verbose',  short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('database', app%database)
        call parser%get('backup',   app%backup)
        call parser%get('vaccum',   app%vacuum)
        call parser%get('wal',      app%wal)
        call parser%get('verbose',  app%verbose)

        rc = validate(app)
    end function read_args

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (dm_file_exists(app%backup)) then
            call dm_error_out(rc, 'backup database exists')
            return
        end if

        if (app%vacuum .and. app%wal) then
            call dm_error_out(rc, 'argument --wal conflicts with --vacuum')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine backup_callback(remaining, page_count)
        !! Prints progess to standard output of SQLite backup API is selected.
        !! The cursor is reset to the first column of the line on each
        !! invokation.
        integer, intent(in) :: remaining  !! Pages remaining.
        integer, intent(in) :: page_count !! Total count of pages.

        write (*, '(a1, "[0GProgress: ", f5.1, " %")', advance='no') &
            ASCII_ESC, 100.0 * (page_count - remaining) / page_count
    end subroutine backup_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_db_version(.true.)
    end subroutine version_callback
end program dmbackup
